FreeBSD

I would like to build dexed on FreeBSD both standalone and as a plugin for DAWs.
Please add instructions on how to build dexed on FreeBSD.

FreeBSD has a `juce-6.0.8_4` package that includes `/usr/local/bin/Projucer`. So I assume this can be used to build dexed?

Getting

```
FreeBSD% Projucer --resave Dexed.jucer

Re-saving file: /usr/home/user/dexed/Dexed.jucer
Error when saving: At least one of your JUCE module paths is invalid!
Please ensure each module path points to the correct JUCE modules folder.
```

Unfortunately it does not say which JUCE module path is invalid. I don't even know what a JUCE module path is or where it is set.

When I run `Projucer` with no arguments, the GUI talks about Linux but not about FreeBSD. Yet I am running it on FreeBSD. What gives?

As a workaround, `sudo ln -s /usr/local/include/X11 /usr/include/` lets me build until this point:

```
[ 61%] Building CXX object Source/CMakeFiles/Dexed.dir/__/libs/JUCE/modules/juce_core/juce_core.cpp.o                                                                                             
In file included from /tmp/dexed/libs/JUCE/modules/juce_core/juce_core.cpp:45:
/usr/include/sys/timeb.h:44:2: error: "this file includes <sys/timeb.h> which is deprecated" [-Werror,-W#warnings]
#warning "this file includes <sys/timeb.h> which is deprecated"
 ^
[ 62%] Building CXX object Source/CMakeFiles/Dexed.dir/__/libs/JUCE/modules/juce_audio_devices/juce_audio_devices.cpp.o                                                                           
[ 63%] Building CXX object Source/CMakeFiles/Dexed.dir/__/libs/JUCE/modules/juce_events/juce_events.cpp.o                                                                                         
1 error generated.
```

Maybe we should use FreeBSD's JUCE from `/usr/local/include/JUCE-6.0.8/` instead?

```
[ 98%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp.o                                                     
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp:26:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/VST3/juce_VST3_Wrapper.cpp:42:
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/juce_VST3Headers.h:27:3: error: To build JUCE VST3 plug-ins on BSD you must use an external BSD-compatible VST3 SDK with JUCE_CUSTOM_VST3_SDK=1
 #error To build JUCE VST3 plug-ins on BSD you must use an external BSD-compatible VST3 SDK with JUCE_CUSTOM_VST3_SDK=1
  ^
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp:26:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/VST3/juce_VST3_Wrapper.cpp:42:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/juce_VST3Headers.h:110:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/base/source/baseiids.cpp:38:
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:222:17: error: expected ';' after top level declarator
int32 PLUGIN_API atomicAdd (int32& value, int32 amount);
                ^
                ;                                                                                
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:2: error: 'virtual' can only appear on non-static member functions
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:28: error: expected ';' at end of declaration list
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
                                  ^
                                  ;                                                              
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:375:2: error: 'virtual' can only appear on non-static member functions
        virtual uint32 PLUGIN_API addRef () = 0;
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:375:17: error: duplicate member 'PLUGIN_API'
        virtual uint32 PLUGIN_API addRef () = 0;
                       ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:18: note: previous declaration is here
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
                        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:375:27: error: expected ';' at end of declaration list
        virtual uint32 PLUGIN_API addRef () = 0;
                                 ^
                                 ;                                                               
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:379:2: error: 'virtual' can only appear on non-static member functions
        virtual uint32 PLUGIN_API release () = 0;
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:379:17: error: duplicate member 'PLUGIN_API'
        virtual uint32 PLUGIN_API release () = 0;
                       ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:18: note: previous declaration is here
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
                        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:379:27: error: expected ';' at end of declaration list
        virtual uint32 PLUGIN_API release () = 0;
                                 ^
                                 ;                                                               
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:493:26: error: no member named 'queryInterface' in 'Steinberg::FUnknown'
        if (unknown && unknown->queryInterface (getTUID<I> (), (void**)&this->ptr) != kResultOk)
                       ~~~~~~~  ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:502:26: error: no member named 'queryInterface' in 'Steinberg::FUnknown'
        if (unknown && unknown->queryInterface (getTUID<I> (), (void**)&newPtr) == kResultOk)
                       ~~~~~~~  ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:544:7: error: no member named 'release' in 'Steinberg::FUnknown'
                        u->release ();
                        ~  ^
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp:26:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/VST3/juce_VST3_Wrapper.cpp:42:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/juce_VST3Headers.h:110:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/base/source/baseiids.cpp:39:
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:35:2: error: 'virtual' can only appear on non-static member functions
        virtual void PLUGIN_API setText (const char8* text) = 0;    
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:35:15: error: field has incomplete type 'void'
        virtual void PLUGIN_API setText (const char8* text) = 0;    
                     ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:35:25: error: expected ';' at end of declaration list
        virtual void PLUGIN_API setText (const char8* text) = 0;    
                               ^
                               ;                                                                 
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:54:2: error: 'virtual' can only appear on non-static member functions
        virtual void PLUGIN_API setText8 (const char8* text) = 0;    
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:54:15: error: field has incomplete type 'void'
        virtual void PLUGIN_API setText8 (const char8* text) = 0;    
                     ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:54:25: error: expected ';' at end of declaration list
        virtual void PLUGIN_API setText8 (const char8* text) = 0;    
                               ^
                               ;                                                                 
fatal error: too many errors emitted, stopping now [-ferror-limit=]                              
20 errors generated.
make[2]: *** [Source/CMakeFiles/Dexed_VST3.dir/build.make:202: Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:463: Source/CMakeFiles/Dexed_VST3.dir/all] Fehler 2
make: *** [Makefile:136: all] Fehler 2

```

With

```
if(CMAKE_SYSTEM_NAME MATCHES "BSD")
target_compile_definitions(${BaseTargetName} PUBLIC
    JUCE_CUSTOM_VST3_SDK=1
)
endif()
```

getting

```
[ 87%] Built target Dexed_Standalone
Consolidate compiler generated dependencies of target Dexed_VST3
[ 88%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_AAX.cpp.o                                                      
[ 89%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_RTAS_1.cpp.o                                                   
[ 90%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_RTAS_2.cpp.o                                                   
[ 91%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_RTAS_3.cpp.o                                                   
[ 92%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_RTAS_4.cpp.o                                                   
[ 94%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_RTAS_utils.cpp.o                                               
[ 95%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_Standalone.cpp.o                                               
[ 96%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_Unity.cpp.o                                                    
[ 97%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST2.cpp.o                                                     
[ 98%] Building CXX object Source/CMakeFiles/Dexed_VST3.dir/__/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp.o                                                     
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp:26:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/VST3/juce_VST3_Wrapper.cpp:42:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/juce_VST3Headers.h:110:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/base/source/baseiids.cpp:38:
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:222:17: error: expected ';' after top level declarator
int32 PLUGIN_API atomicAdd (int32& value, int32 amount);
                ^
                ;                                                                                
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:2: error: 'virtual' can only appear on non-static member functions
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:28: error: expected ';' at end of declaration list
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
                                  ^
                                  ;                                                              
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:375:2: error: 'virtual' can only appear on non-static member functions
        virtual uint32 PLUGIN_API addRef () = 0;
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:375:17: error: duplicate member 'PLUGIN_API'
        virtual uint32 PLUGIN_API addRef () = 0;
                       ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:18: note: previous declaration is here
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
                        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:375:27: error: expected ';' at end of declaration list
        virtual uint32 PLUGIN_API addRef () = 0;
                                 ^
                                 ;                                                               
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:379:2: error: 'virtual' can only appear on non-static member functions
        virtual uint32 PLUGIN_API release () = 0;
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:379:17: error: duplicate member 'PLUGIN_API'
        virtual uint32 PLUGIN_API release () = 0;
                       ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:370:18: note: previous declaration is here
        virtual tresult PLUGIN_API queryInterface (const TUID _iid, void** obj) = 0;
                        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:379:27: error: expected ';' at end of declaration list
        virtual uint32 PLUGIN_API release () = 0;
                                 ^
                                 ;                                                               
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:493:26: error: no member named 'queryInterface' in 'Steinberg::FUnknown'
        if (unknown && unknown->queryInterface (getTUID<I> (), (void**)&this->ptr) != kResultOk)
                       ~~~~~~~  ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:502:26: error: no member named 'queryInterface' in 'Steinberg::FUnknown'
        if (unknown && unknown->queryInterface (getTUID<I> (), (void**)&newPtr) == kResultOk)
                       ~~~~~~~  ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/funknown.h:544:7: error: no member named 'release' in 'Steinberg::FUnknown'
                        u->release ();
                        ~  ^
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp:26:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_plugin_client/VST3/juce_VST3_Wrapper.cpp:42:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/juce_VST3Headers.h:110:
In file included from /tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/base/source/baseiids.cpp:39:
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:35:2: error: 'virtual' can only appear on non-static member functions
        virtual void PLUGIN_API setText (const char8* text) = 0;    
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:35:15: error: field has incomplete type 'void'
        virtual void PLUGIN_API setText (const char8* text) = 0;    
                     ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:35:25: error: expected ';' at end of declaration list
        virtual void PLUGIN_API setText (const char8* text) = 0;    
                               ^
                               ;                                                                 
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:54:2: error: 'virtual' can only appear on non-static member functions
        virtual void PLUGIN_API setText8 (const char8* text) = 0;    
        ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:54:15: error: field has incomplete type 'void'
        virtual void PLUGIN_API setText8 (const char8* text) = 0;    
                     ^
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:54:25: error: expected ';' at end of declaration list
        virtual void PLUGIN_API setText8 (const char8* text) = 0;    
                               ^
                               ;                                                                 
/tmp/dexed/libs/JUCE/modules/juce_audio_processors/format_types/VST3_SDK/pluginterfaces/base/istringresult.h:56:2: error: 'virtual' can only appear on non-static member functions
        virtual void PLUGIN_API setText16 (const char16* text) = 0;
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]                              
20 errors generated.
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -56,7 +56,7 @@ target_compile_options(${PROJECT_NAME} PUBLIC
         )
 elseif(UNIX AND NOT APPLE)
 target_compile_options(${PROJECT_NAME} PUBLIC
-        -Werror
+        #-Werror
         -Wno-deprecated-declarations
         -Wno-unused-value
         )

--- a/Source/CMakeLists.txt
+++ b/Source/CMakeLists.txt
@@ -1,5 +1,17 @@
 set(BaseTargetName Dexed)
 
+set(DEXED_JUCE_FORMATS Standalone)
+
+if(NOT DEXED_SKIP_VST3)
+  list(APPEND DEXED_JUCE_FORMATS VST3)
+endif()
+
+if (APPLE)
+  list(APPEND DEXED_JUCE_FORMATS AU)
+endif()
+
+message(STATUS "Building Dexed in formats: ${DEXED_JUCE_FORMATS}")
+
 juce_add_plugin("${BaseTargetName}"
         # VERSION ...                               # Set this if the plugin version is different to the project version
         ICON_BIG "../Resources/ui/dexedIcon.png"
@@ -13,7 +25,7 @@ juce_add_plugin("${BaseTargetName}"
         COPY_PLUGIN_AFTER_BUILD FALSE
         PLUGIN_MANUFACTURER_CODE DGSB
         PLUGIN_CODE Dexd
-        FORMATS AU VST3 Standalone
+        FORMATS ${DEXED_JUCE_FORMATS}
         PRODUCT_NAME "Dexed"
         DESCRIPTION "Dexed FM Synth"
 )
@@ -54,11 +66,19 @@ target_sources(${BaseTargetName} PRIVATE
         msfa/tuning.cc
 )
 
+if(DEXED_NO_ALSA)
+    set(DEXED_ALSA FALSE)
+    message(STATUS "Building Dexed without ALSA support")
+else()
+    set(DEXED_ALSA TRUE)
+endif()
+
 target_compile_definitions(${BaseTargetName} PUBLIC
     JUCE_WEB_BROWSER=0
     JUCE_USE_CURL=0
     JUCE_VST3_CAN_REPLACE_VST2=0
     JUCE_JACK=1
+    JUCE_ALSA=$<IF:$<BOOL:${DEXED_ALSA}>,1,0>
     JUCE_MODAL_LOOPS_PERMITTED=1 # needed for FileBrowser in CartManager
     JUCE_DISPLAY_SPLASH_SCREEN=0
 )

